cli/src/errors/error.rs

```diff
@@ -1,49 +1,92 @@
-#[cfg(backtrace)]
-use std::backtrace::Backtrace;
 use std::convert::From;
+use std::error::Error;
+use std::fmt;
 
 use anyhow::Result;
 use log::error;
-use thiserror::Error;
 
+use crate::errors::ErrorDetails;
 use crate::errors::ExitCode;
 
-#[derive(Error, Debug)]
-#[error("{msg}")]
+#[derive(Debug)]
 pub struct ApolloError {
-    msg: String,
-    #[cfg(backtrace)]
-    backtrace: Backtrace,
-    /// The result of `error.exit_code()`.
-    exit_code: ExitCode,
+    inner: Box<Inner>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    kind: ErrorDetails,
+    source: Option<Box<dyn Error>>,
 }
 
 #[allow(dead_code)]
 impl ApolloError {
     /// Returns the process exit code that should be returned if the process exits with this error.
     pub fn exit_code(&self) -> ExitCode {
-        self.exit_code
+        self.inner.kind.exit_code()
+    }
+
+    pub fn from_source<E>(source: E, kind: ErrorDetails) -> Self
+    where
+        E: Into<Box<dyn Error>>,
+    {
+        ApolloError {
+            inner: Box::new(Inner {
+                kind,
+                source: Some(source.into()),
+            }),
+        }
+    }
+
+    pub fn kind(&self) -> &ErrorDetails {
+        &self.inner.kind
     }
 }
 
-/// The failure trait for all Apollo errors.
-pub trait ApolloFail: std::error::Error {
-    /// Returns the process exit code that should be returned if the process exits with this error.
-    fn exit_code(&self) -> ExitCode;
+impl fmt::Display for ApolloError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.inner.kind.fmt(f)
+    }
 }
 
-impl<T: ApolloFail> From<T> for ApolloError {
-    fn from(error: T) -> Self {
-        let exit_code = error.exit_code();
+impl Error for ApolloError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.inner.source.as_ref().map(|b| b.as_ref())
+    }
+}
+
+impl From<ErrorDetails> for ApolloError {
+    fn from(kind: ErrorDetails) -> Self {
         ApolloError {
-            msg: error.to_string(),
-            exit_code,
+            inner: Box::new(Inner { kind, source: None }),
         }
     }
 }
 
+pub trait Context<T> {
+    fn with_context<F>(self, f: F) -> Fallible<T>
+    where
+        F: FnOnce() -> ErrorDetails;
+}
+
+impl<T, E> Context<T> for Result<T, E>
+where
+    E: Error + 'static,
+{
+    fn with_context<F>(self, f: F) -> Fallible<T>
```

Comment:
Q: When does one use `self` vs `&self`?

Reply:
`self` takes ownership over the object which we want to do since we are changing the underlying error. 
![image](https://user-images.githubusercontent.com/4967600/82965268-a37c5780-9f95-11ea-8a06-4b9c02c90abc.png)